iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/cache/ScanPlanCacheKey.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.gravitino.iceberg.service.cache;
+
+import com.google.common.base.Objects;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.rest.requests.PlanTableScanRequest;
+
+/** Cache key for Iceberg table scan plans. */
+public class ScanPlanCacheKey {
+  private final TableIdentifier tableIdentifier;
+  private final Long snapshotId;
+  private final Long startSnapshotId;
+  private final Long endSnapshotId;
+  private final Expression filter; // Store the filter object instead of string
+  private final String selectStr;
+  private final String statsFieldsStr;
+  private final boolean caseSensitive;
+  private final boolean useSnapshotSchema;
+
+  private ScanPlanCacheKey(
+      TableIdentifier tableIdentifier,
+      Long snapshotId,
+      Long startSnapshotId,
+      Long endSnapshotId,
+      Expression filter,
+      String select,
+      String statsFields,
+      boolean caseSensitive,
+      boolean useSnapshotSchema) {
+    this.tableIdentifier = tableIdentifier;
+    this.snapshotId = snapshotId;
+    this.startSnapshotId = startSnapshotId;
+    this.endSnapshotId = endSnapshotId;
+    this.filter = filter;
+    this.selectStr = select;
+    this.statsFieldsStr = statsFields;
+    this.caseSensitive = caseSensitive;
+    this.useSnapshotSchema = useSnapshotSchema;
+  }
+
+  /**
+   * Creates a cache key from table identifier, table, and scan request.
+   *
+   * @param tableIdentifier the table identifier
+   * @param table the Iceberg table
+   * @param scanRequest the scan request containing filters and projections
+   * @return a new cache key
+   */
+  public static ScanPlanCacheKey create(
+      TableIdentifier tableIdentifier, Table table, PlanTableScanRequest scanRequest) {
+
+    // Use current snapshot if not specified
+    Long snapshotId = scanRequest.snapshotId();
+    if (snapshotId == null && table.currentSnapshot() != null) {
+      snapshotId = table.currentSnapshot().snapshotId();
+    }
+
+    // Include startSnapshotId and endSnapshotId in the key
+    Long startSnapshotId = scanRequest.startSnapshotId();
+    Long endSnapshotId = scanRequest.endSnapshotId();
+
+    // Store the filter expression object directly instead of its string representation.
+    // This approach relies on Expression's equals() and hashCode() implementations,
+    // which are designed to correctly handle semantic equality of expressions.
+    Expression filter = scanRequest.filter();
```

Comment:
Could you add test for this?  seems Expression provides an `isEquivalentTo` method and doesn't override `equals` and `hashCode` methods.